platform/cog-platform-gtk4.c

```diff
@@ -0,0 +1,651 @@
+}
+
+static gboolean action_quit(GtkWidget* widget, GVariant* args,
+    gpointer user_data)
+{
+    if (g_application_get_default())
+        g_application_quit(g_application_get_default());
+
+    return TRUE;
+}
+
+static gboolean action_activate_entry(GtkWidget* widget, GVariant* args,
+    gpointer user_data)
+{
+    gtk_widget_grab_focus(win.url_entry);
+    return TRUE;
+}
+
+static void setup_window(struct platform_window* window)
+{
+    g_assert_nonnull(window);
+    window->gtk_window = gtk_window_new();
+    g_object_set(window->gtk_window, "default-width", DEFAULT_WIDTH,
+        "default-height", DEFAULT_HEIGHT, NULL);
+    g_signal_connect(window->gtk_window, "destroy", G_CALLBACK(quit_cb), NULL);
+
+    GtkWidget* header_bar = gtk_header_bar_new();
+    GtkWidget* left_stack = gtk_stack_new();
+    GtkWidget* buttons_box = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 0);
+
+    window->back_button = gtk_button_new_from_icon_name("go-previous-symbolic");
+    g_signal_connect(window->back_button, "clicked", G_CALLBACK(on_back_clicked),
+        NULL);
+    gtk_box_append(GTK_BOX(buttons_box), window->back_button);
+
+    window->forward_button = gtk_button_new_from_icon_name("go-next-symbolic");
+    g_signal_connect(window->forward_button, "clicked",
+        G_CALLBACK(on_forward_clicked), NULL);
+    gtk_box_append(GTK_BOX(buttons_box), window->forward_button);
+
+    GtkWidget* refresh_button = gtk_button_new_from_icon_name("view-refresh-symbolic");
+    g_signal_connect(refresh_button, "clicked", G_CALLBACK(on_refresh_clicked),
+        NULL);
+    gtk_box_append(GTK_BOX(buttons_box), refresh_button);
+
+    gtk_stack_add_named(GTK_STACK(left_stack), buttons_box, "buttons");
+    gtk_header_bar_pack_start(GTK_HEADER_BAR(header_bar), left_stack);
+
+    GtkWidget* title_box = gtk_box_new(GTK_ORIENTATION_VERTICAL, 0);
+    gtk_widget_set_valign(title_box, GTK_ALIGN_CENTER);
+    gtk_widget_set_hexpand(title_box, TRUE);
+
+    window->url_entry = gtk_entry_new();
+    g_signal_connect(window->url_entry, "activate",
+        G_CALLBACK(on_entry_activated), NULL);
+    gtk_widget_set_parent(window->url_entry, title_box);
+
+    gtk_header_bar_set_title_widget(GTK_HEADER_BAR(header_bar), title_box);
+    gtk_header_bar_set_show_title_buttons(GTK_HEADER_BAR(header_bar), TRUE);
+    gtk_window_set_titlebar(GTK_WINDOW(window->gtk_window), header_bar);
+
+    window->gl_drawing_area = gtk_gl_area_new();
+    gtk_widget_set_hexpand(window->gl_drawing_area, TRUE);
+    gtk_widget_set_vexpand(window->gl_drawing_area, TRUE);
+    g_object_set(window->gl_drawing_area, "use-es", TRUE, NULL);
+    g_signal_connect(window->gl_drawing_area, "realize", G_CALLBACK(realize),
+        NULL);
+    g_signal_connect(window->gl_drawing_area, "render", G_CALLBACK(render), NULL);
+    g_signal_connect(window->gl_drawing_area, "resize", G_CALLBACK(resize), NULL);
+
+    GtkGesture* press = gtk_gesture_click_new();
```

Comment:
This will leak the `GtkGesture`, because the `_new()` functions for event controllers return a full reference, not a floating one (the parent class `GtkEventController` does _not_ inherit from `GInitiallyUnowned`).

Using `g_autoptr` is probably the easiest way to go:

 ```suggestion
    g_autoptr(GtkGesture) press = gtk_gesture_click_new();
```

The same goes for the rest of event controllers below.

Reply:
Is it really leaking? I thought the ownership would transfer to the widget when calling `gtk_widget_add_controller()`.

Reply:
Ouch, you're right! :facepalm: 

Somehow browsing the GTK4 documentation locally with Devhelp is not showing me notes like “transfer full” in the descriptions of function parameters—the documentation at [the GNOME documentation web site](https://developer.gnome.org/gtk4/stable/GtkWidget.html#gtk-widget-add-controller) does show them.

(Sorry about the noise, I have to figure out why Devhelp is being odd here, it happens only with the GTK4 docs!)